### PR TITLE
Use uint8_t instead of unsigned char in crypto APIs

### LINKED ISSUE #2402: Change Crypto APIs to use uint8_t* instead of "unsigned char*"
#### Problem
The sizeof(char) may differ based on compiler/build configuration. Use of this type makes the crypto API ambiguous.

 #### Proposed Solution
Use pointers to uint8_t type. Assert and cast inside `src/crypto` as needed.

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
The sizeof(char) may differ based on compiler/build configuration. Use of this type makes the crypto API ambiguous.

 #### Summary of Changes
Use pointers to uint8_t type. Assert and cast inside src/crypto as needed.

fixes #2402

### CODE REVIEW

File: src/crypto/tests/PBKDF2_SHA256_test_vectors.h [hunk 1]
```diff
@@ -28,71 +28,71 @@
...
 };
 
-static const unsigned char chiptest_key1[]                    = { 0x12, 0x0f, 0xb6, 0xcf, 0xfc, 0xf8, 0xb3, 0x2c, 0x43, 0xe7,
+static const uint8_t chiptest_key1[]                    = { 0x12, 0x0f, 0xb6, 0xcf, 0xfc, 0xf8, 0xb3, 0x2c, 0x43, 0xe7,
                                                0x22, 0x52, 0x56, 0xc4, 0xf8, 0x37, 0xa8, 0x65, 0x48, 0xc9 };
-static const struct pbkdf2_test_vector chiptest_test_vector_1 = { .password = (const unsigned char *) "password",
+static const struct pbkdf2_test_vector chiptest_test_vector_1 = { .password = (const uint8_t *) "password",
                                                                   .plen     = 8,
-                                                                  .salt     = (const unsigned char *) "salt",
+                                                                  .salt     = (const uint8_t *) "salt",
```
Comment (gerickson) [inline src/crypto/tests/PBKDF2_SHA256_test_vectors.h:46]: Prefer C++-style casts where possible.

File: src/crypto/tests/PBKDF2_SHA256_test_vectors.h [hunk 2]
```diff
@@ -28,71 +28,71 @@
...
+    const uint8_t * key;
     unsigned tcId;
     int result;
 };
 
-static const unsigned char chiptest_key1[]                    = { 0x12, 0x0f, 0xb6, 0xcf, 0xfc, 0xf8, 0xb3, 0x2c, 0x43, 0xe7,
+static const uint8_t chiptest_key1[]                    = { 0x12, 0x0f, 0xb6, 0xcf, 0xfc, 0xf8, 0xb3, 0x2c, 0x43, 0xe7,
                                                0x22, 0x52, 0x56, 0xc4, 0xf8, 0x37, 0xa8, 0x65, 0x48, 0xc9 };
-static const struct pbkdf2_test_vector chiptest_test_vector_1 = { .password = (const unsigned char *) "password",
+static const struct pbkdf2_test_vector chiptest_test_vector_1 = { .password = (const uint8_t *) "password",
```
Comment (gerickson) [inline src/crypto/tests/PBKDF2_SHA256_test_vectors.h:44]: Prefer C++-style casts where possible.

File: src/crypto/tests/PBKDF2_SHA256_test_vectors.h [hunk 3]
```diff
@@ -28,71 +28,71 @@
...
                                                                   .key_len  = 20,
                                                                   .key      = chiptest_key1,
                                                                   .tcId     = 1,
                                                                   .result   = CHIP_NO_ERROR };
 
-static const unsigned char chiptest_key2[]                    = { 0xae, 0x4d, 0x0c, 0x95, 0xaf, 0x6b, 0x46, 0xd3, 0x2d, 0x0a,
+static const uint8_t chiptest_key2[]                    = { 0xae, 0x4d, 0x0c, 0x95, 0xaf, 0x6b, 0x46, 0xd3, 0x2d, 0x0a,
                                                0xdf, 0xf9, 0x28, 0xf0, 0x6d, 0xd0, 0x2a, 0x30, 0x3f, 0x8e };
-static const struct pbkdf2_test_vector chiptest_test_vector_2 = { .password = (const unsigned char *) "password",
+static const struct pbkdf2_test_vector chiptest_test_vector_2 = { .password = (const uint8_t *) "password",
```
Comment (gerickson) [inline src/crypto/tests/PBKDF2_SHA256_test_vectors.h:56]: Prefer C++-style casts where possible.

File: src/crypto/tests/PBKDF2_SHA256_test_vectors.h [hunk 4]
```diff
@@ -28,71 +28,71 @@
...
                                                                   .plen     = 8,
-                                                                  .salt     = (const unsigned char *) "salt",
+                                                                  .salt     = (const uint8_t *) "salt",
```
Comment (gerickson) [inline src/crypto/tests/PBKDF2_SHA256_test_vectors.h:58]: Prefer C++-style casts where possible.

File: src/transport/SecurePairingSession.cpp [hunk 5]
```diff
@@ -62,10 +62,10 @@ CHIP_ERROR SecurePairingSession::Init(uint32_t setupCode, uint32_t pbkdf2IterCou
     VerifyOrExit(saltLen > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrExit(delegate != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
 
-    err = mSpake2p.Init((const unsigned char *) kSpake2pContext, strlen(kSpake2pContext));
+    err = mSpake2p.Init((const uint8_t *) kSpake2pContext, strlen(kSpake2pContext));
```
Comment (gerickson) [inline src/transport/SecurePairingSession.cpp:65]: Prefer C++-style casts where possible.

File: src/transport/SecurePairingSession.cpp [hunk 6]
```diff
@@ -62,10 +62,10 @@ CHIP_ERROR SecurePairingSession::Init(uint32_t setupCode, uint32_t pbkdf2IterCou
...
     SuccessOrExit(err);
```
Comment (gerickson) [inline src/transport/SecurePairingSession.cpp:67]: Prefer C++-style casts where possible.

File: src/transport/SecurePairingSession.cpp [hunk 7]
```diff
@@ -149,7 +149,7 @@ CHIP_ERROR SecurePairingSession::Pair(uint32_t peerSetUpPINCode, uint32_t pbkdf2
     CHIP_ERROR err = Init(peerSetUpPINCode, pbkdf2IterCount, salt, saltLen, myNodeId, myKeyId, delegate);
     SuccessOrExit(err);
 
-    err = mSpake2p.BeginProver((const unsigned char *) "", 0, (const unsigned char *) "", 0, &mWS[0][0], kSpake2p_WS_Length,
+    err = mSpake2p.BeginProver((const uint8_t *) "", 0, (const uint8_t *) "", 0, &mWS[0][0], kSpake2p_WS_Length,
```
Comment (gerickson) [inline src/transport/SecurePairingSession.cpp]: Prefer C++-style casts where possible.

File: src/transport/SecurePairingSession.cpp [hunk 8]
```diff
@@ -62,10 +62,10 @@ CHIP_ERROR SecurePairingSession::Init(uint32_t setupCode, uint32_t pbkdf2IterCou
     VerifyOrExit(saltLen > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
     VerifyOrExit(delegate != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
 
-    err = mSpake2p.Init((const unsigned char *) kSpake2pContext, strlen(kSpake2pContext));
+    err = mSpake2p.Init((const uint8_t *) kSpake2pContext, strlen(kSpake2pContext));
     SuccessOrExit(err);
 
-    err = pbkdf2_sha256((const unsigned char *) &setupCode, sizeof(setupCode), salt, saltLen, pbkdf2IterCount, sizeof(mWS),
+    err = pbkdf2_sha256((const uint8_t *) &setupCode, sizeof(setupCode), salt, saltLen, pbkdf2IterCount, sizeof(mWS),
```
Comment (gerickson) [inline src/transport/SecurePairingSession.cpp]: Prefer C++-style casts where possible.

File: src/transport/SecurePairingSession.cpp [hunk 9]
```diff
@@ -219,7 +219,7 @@ CHIP_ERROR SecurePairingSession::HandleCompute_pA(const MessageHeader & header,
     VerifyOrExit(buf != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
     VerifyOrExit(buf_len == kMAX_Point_Length, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
 
-    err = mSpake2p.BeginVerifier((const unsigned char *) "", 0, (const unsigned char *) "", 0, &mWS[0][0], kSpake2p_WS_Length,
+    err = mSpake2p.BeginVerifier((const uint8_t *) "", 0, (const uint8_t *) "", 0, &mWS[0][0], kSpake2p_WS_Length,
```
Comment (gerickson) [inline src/transport/SecurePairingSession.cpp]: Prefer C++-style casts where possible.

File: src/transport/SecurePairingSession.h [hunk 10]
```diff
@@ -199,24 +199,24 @@ class SecurePairingUsingTestSecret : public SecurePairingSession
...
     }
 
-    CHIP_ERROR DeriveSecureSession(const unsigned char * info, size_t info_len, SecureSession & session)
+    CHIP_ERROR DeriveSecureSession(const uint8_t * info, size_t info_len, SecureSession & session)
     {
         const char * secret = "Test secret for key derivation";
         size_t secretLen    = strlen(secret);
-        return session.InitFromSecret((const unsigned char *) secret, secretLen, (const unsigned char *) "", 0,
-                                      (const unsigned char *) secret, secretLen);
+        return session.InitFromSecret((const uint8_t *) secret, secretLen, (const uint8_t *) "", 0,
```
Comment (gerickson) [inline src/transport/SecurePairingSession.h]: Prefer C++-style casts where possible.

File: src/transport/SecureSession.cpp [hunk 11]
```diff
@@ -166,8 +166,8 @@ CHIP_ERROR SecureSession::Encrypt(const unsigned char * input, size_t input_leng
     error = GetAdditionalAuthData(header, AAD, aadLen);
     SuccessOrExit(error);
 
-    error = AES_CCM_encrypt(input, input_length, AAD, aadLen, mKey, sizeof(mKey), (const unsigned char *) IV, sizeof(IV), output,
-                            (unsigned char *) &tag, sizeof(tag));
+    error = AES_CCM_encrypt(input, input_length, AAD, aadLen, mKey, sizeof(mKey), IV, sizeof(IV), output,
```
Comment (gerickson) [inline src/transport/SecureSession.cpp]: Prefer C++-style casts where possible.

File: src/transport/SecureSessionMgr.cpp [hunk 12]
```diff
@@ -187,7 +187,7 @@ CHIP_ERROR SecureSessionMgrBase::NewPairing(Optional<NodeId> peerNodeId, const O
 
     if (state != nullptr)
     {
-        err = pairing->DeriveSecureSession((const unsigned char *) kSpake2pI2RSessionInfo, strlen(kSpake2pI2RSessionInfo),
+        err = pairing->DeriveSecureSession((const uint8_t *) kSpake2pI2RSessionInfo, strlen(kSpake2pI2RSessionInfo),
```
Comment (gerickson) [inline src/transport/SecureSessionMgr.cpp:190]: Prefer C++-style casts where possible.

File: src/transport/tests/TestSecurePairingSession.cpp [hunk 13]
```diff
@@ -76,10 +76,10 @@ void SecurePairingWaitTest(nlTestSuite * inSuite, void * inContext)
                    pairing.WaitForPairing(1234, 500, nullptr, 0, Optional<NodeId>::Value(1), 0, &delegate) ==
                        CHIP_ERROR_INVALID_ARGUMENT);
     NL_TEST_ASSERT(inSuite,
-                   pairing.WaitForPairing(1234, 500, (const unsigned char *) "salt", 4, Optional<NodeId>::Value(1), 0, nullptr) ==
+                   pairing.WaitForPairing(1234, 500, (const uint8_t *) "salt", 4, Optional<NodeId>::Value(1), 0, nullptr) ==
```
Comment (gerickson) [inline src/transport/tests/TestSecurePairingSession.cpp:79]: Prefer C++-style casts where possible.

File: src/transport/tests/TestSecureSession.cpp [hunk 14]
```diff
@@ -146,19 +146,19 @@ void SecureChannelDecryptTest(nlTestSuite * inSuite, void * inContext)
...
-    unsigned char output[128];
+    uint8_t output[128];
     // Uninitialized channel
     NL_TEST_ASSERT(inSuite,
                    channel2.Decrypt(encrypted, sizeof(plain_text), output, header) == CHIP_ERROR_INVALID_USE_OF_SESSION_KEY);
     NL_TEST_ASSERT(inSuite,
                    channel2.Init(secure_channel_test_public_key2, sizeof(secure_channel_test_public_key2),
                                  secure_channel_test_private_key1, sizeof(secure_channel_test_private_key1),
-                                 (const unsigned char *) salt, sizeof(salt), (const unsigned char *) info,
+                                 (const uint8_t *) salt, sizeof(salt), (const uint8_t *) info,
```
Comment (gerickson) [inline src/transport/tests/TestSecureSession.cpp]: Prefer C++-style casts where possible.

**Review by gerickson — COMMENTED** [review]
Overall, LGTM. Prefer C++-style casts where possible.

**Comment by bzbarsky-apple** [comment]
>  sizeof(char) may differ based on compiler/build configuration

No, sizeof(char) is always 1, by definition.  What can differ is how many bits a given sizeof value means.

File: src/crypto/CHIPCryptoPAL.cpp [hunk 15]
```diff
@@ -141,19 +141,19 @@ CHIP_ERROR Spake2p::BeginVerifier(const unsigned char * my_identity, size_t my_i
     error = PointLoad(Lin, Lin_len, L);
     VerifyOrExit(error == CHIP_NO_ERROR, error = CHIP_ERROR_INTERNAL);
 
-    state = CHIP_SPAKE2P_STATE_STARTED;
-    role  = CHIP_SPAKE2P_VERIFIER;
+    state = CHIP_SPAKE2P_STATE::CHIP_SPAKE2P_STATE_STARTED;
+    role  = CHIP_SPAKE2P_ROLE::CHIP_SPAKE2P_VERIFIER;
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.cpp:145]: Again, feels like values could be shorter now.  Again, followup, probably separate from the state one.

File: src/crypto/CHIPCryptoPAL.h [hunk 16]
```diff
@@ -71,7 +71,7 @@ const unsigned char spake2p_N_p256[65] = {
 /**
  * Spake2+ state machine to ensure proper execution of the protocol.
  */
-enum
+enum class CHIP_SPAKE2P_STATE
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.h:74]: Do we want to throw in `: uint8_t` at the end to ensure that it stays one-byte as before and the compiler doesn't randomly widen it?

File: src/crypto/CHIPCryptoPAL.h [hunk 17]
```diff
@@ -80,17 +80,15 @@ enum
     CHIP_SPAKE2P_STATE_R2,          // Round two complete
     CHIP_SPAKE2P_STATE_KC,          // Key confirmation complete
 };
-typedef unsigned char CHIP_SPAKE2P_STATE;
 
 /**
  * Spake2+ role.
  */
-enum
+enum class CHIP_SPAKE2P_ROLE
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.h:87]: Again, do we want `: uint8_t`?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Comment by pan-apple** [comment]
@saurabhst, @jelderton, @BroderickCarlin, do you have any comments?

**Review by saurabhst — APPROVED** [review]

**Pull request merged**